io/shared/src/main/scala/scalaz/effect/IO.scala

```diff
@@ -486,6 +474,15 @@ sealed abstract class IO[E, A] { self =>
 }
 
 object IO {
+
+  type MaybeIO[E, A] = (IO[E, A], A => IO[E, A]) => IO[E, A]
+  object MaybeIO {
+    def empty[E, A]: MaybeIO[E, A] = (z, _) => z
+    def just[E, A](a: A): MaybeIO[E, A] = (_, f) => f(a)
+  }
+
+  final case class FromEA[E, A](err: E => IO[E, A], succ: A => IO[E, A])
```

Comment:
I'd also delete this in favor of the `Disjunction` trait I suggested and that way of encoding it.